samcli/lib/test_runner/test_runner_template_generator.py

```diff
@@ -0,0 +1,255 @@
+            LOG.info("Matched ARN `%s` to IAM actions %s", resource_arn, mapping["Action"])
+            return mapping
+
+    # If the supplied ARN is for a resource without any default actions supported
+    return None
+
+
+def _create_iam_statment_string(resource_arn: str) -> str:
+    """
+    Returns an IAM Statement in the form of a YAML string corresponding to the supplied resource ARN.
+
+    The generated Actions are commented out to establish a barrier of consent, preventing customers from accidentally granting permissions that they did not mean to.
+
+    The Action list is empty if there are no IAM Action permissions to generate for the given resource.
+
+    Parameters
+    ----------
+    resource_arn : str
+        The arn of the resource for which the IAM statement is generated.
+
+    Returns
+    -------
+    str:
+        An IAM Statement in the form of a YAML string.
+    """
+
+    # NOTE: The spacing within these strings is needed to keep the resulting YAML indentation valid
+    new_statement_template = (
+        "- Effect: Allow\n"
+        "  Action:\n"
+        "   {%- for action in action_list %}\n"
+        "     # - {{action}}\n"
+        "   {%- endfor %}\n"
+        "  Resource: {{arn}}\n"
+    )
+    mapping = _get_action_mapping(resource_arn)
+
+    if mapping is None:
+        return jinja2.Template(new_statement_template).render(action_list=[], arn=resource_arn).rstrip()
+
+    if mapping["Resource"] in ("AWS::ApiGateway::Api", "AWS::ApiGateway::RestApi"):
+        apiId = _extract_api_id_from_arn(resource_arn)
+        execute_api_arn = f"!Sub arn:${{AWS::Partition}}:execute-api:${{AWS::Region}}:${{AWS::AccountId}}:{apiId}/<STAGE>/GET/<RESOURCE_PATH>"
+        return (
+            jinja2.Template(new_statement_template).render(action_list=mapping["Action"], arn=execute_api_arn).rstrip()
+        )
+
+    return jinja2.Template(new_statement_template).render(action_list=mapping["Action"], arn=resource_arn).rstrip()
+
+
+def _query_tagging_api(tag_filters: dict) -> Union[dict, None]:
+    """
+    Queries the Tagging API to retrieve the ARNs of every resource with the given tags.
+
+    NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#get-resources
+
+    Parameters
+    ----------
+    tag_filters : dict
+        The tag filters to restrict output to only those resources with the given tags.
+
+        NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#options
+
+    Returns
+    -------
+    dict
+        A ResourceTagMappingList, which contains a list of resource ARNs and tags associated with each.
+
+        NOTE: https://docs.aws.amazon.com/cli/latest/reference/resourcegroupstaggingapi/get-resources.html#output
+
+
+    """
+    try:
+        return boto3.client("resourcegroupstaggingapi").get_resources(TagFilters=tag_filters)["ResourceTagMappingList"]
```

Comment:
I would suggest using `get_boto_client_provider_with_config` which will return a client with `region` and `profile`  settings as well as updating user agent with SAM CLI.

That function will return a function reference which can be used later on, so that once you have `region` and `profile` settings, you can initalize the function reference, and you can use it later to initialize clients. 
```
# initalize the client provider
boto_client_provider = get_boto_client_provider_with_config(region=region, profile=profile)

# use it later to initialize the client
resource_group_tagging_api_client = boto_client_provider("resourcegroupstaggingapi")
```

Reply:
What are the advantages of this? Why does the client need `region` and `profile` settings? 

I'm also not sure what `updating user agent with SAM CLI` means

Reply:
Edit: Andrew mentioned that it marks requests as coming from the SAM CLI instead of just boto3. I think this is good to have for consistency

Reply:
If we will tie this to a command, then users might run their command with specific region and profile settings. We usually use `default` profile which is used by default as the name suggest. But if user invoke their command with another profile name using default might fail since there might not be a default profile or their default profile might not have sufficient permission to execute this.